lib/icinga/downtime.cpp

```diff
@@ -427,6 +443,26 @@ bool Downtime::CanBeTriggered()
 	return true;
 }
 
+void Downtime::SetupCleanupTimer()
+{
+	if (!m_CleanupTimer) {
+		m_CleanupTimer = Timer::Create();
+
+		auto name (GetName());
+
+		m_CleanupTimer->OnTimerExpired.connect([name=std::move(name)](const Timer * const&) {
+			auto downtime (Downtime::GetByName(name));
+
+			if (downtime && downtime->IsExpired()) {
+				RemoveDowntime(name, false, false, true);
+			}
+		});
+	}
+
+	m_CleanupTimer->Reschedule((GetFixed() || !IsTriggered() ? GetEndTime() : GetTriggerTime() + GetDuration()) + 0.1);
+	m_CleanupTimer->Start();
```

Comment:
With the call from `TriggerDowntime()`, this can be called for an already started timer. It doesn't look like `Timer::Start()` would handle this correctly as it unconditionally increments a counter `l_AliveTimers`, i.e. it may count these timers multiple times (but the corresponding decrement operation would only subtract one): https://github.com/Icinga/icinga2/blob/1665cc925e3e29b40c8b49fa0ef1cd35f437cc7e/lib/base/timer.cpp#L158-L160